contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,525 @@
+
+The above table is an evidence that we need server-side validation as well, next to the client-side validation we easily get, nearly for free, by serving Swagger/OpenAPI specs in apiextension-apiserver. 
+
+This is especially true in situations when CRs are used by components that are out of the control of the admin. Example: A user can create a database CR for a Database-As-A-Service. In this case, only server-side validation can give confidence that the CRs are well formed.
+
+### Existing Instances and changing the Schema
+
+If the schema is made stricter later, the existing CustomResources might no longer comply with the spec. This will make them unchangeable and essentially read-only.
+
+To avoid this, it is the responsibility of the user to make sure that any changes made to the schema are such that the existing CustomResources remain validated. 
+ 
+Note:
+
+1. This is the same behavior that we require for native resources. Validation cannot be made stricter in later Kubernetes versions without breaking compatibility.
+
+2. For migration of CRDs with no validation to CRDs with validation, we can create a controller that will validate and annotate invalid CRs once the spec changes, so that the custom controller can choose to delete them (this is also essentially the status condition of the CRD). This can be achieved, but it is not part of the proposal.
+
+### Outlook to Status Sub-Resources
+
+As another most-wanted feature, a Status sub-resource might be proposed and implemented for CRDs. The JSON-Schema proposed here might as well cover the Status field of a CR. For now this is not handled or validated in a particular way.
+ 
+When the Status sub-resource exists some day, the /status endpoint will receive a full CR object, but only the status field is to be validated. We propose to enforce the JSON-Schema structure to be of the shape:
+
+```json
+{"type":"object", "properties":{"status": ..., "a": ..., "b": ...}}
+```
+Then we can validate the status against the sub-schema easily. Hence, this proposal will be compatible with a later sub-resource extension.
+
+### Outlook Admission Webhook
+
+Apiextensions-apiserver uses the normal REST endpoint implementation and only customizes the registry and the codecs. The admission plugins are inherited from the kube-apiserver (when running inside of it via apiserver delegation) and therefore they are supposed to apply to CRs as well.
+
+It is [verified](https://github.com/kubernetes/kubernetes/pull/47252) that CRDs work well with initializers. It is also expected that webhook admission prototyped at https://github.com/kubernetes/kubernetes/pull/46316 will work with CRs out of the box. Hence, for more advanced validation webhook admission is an option as well (when it is merged).
+
+Clearly, providing such a webhook is more complex and therefore cannot replace a built-in JSON-Schema validation and its use-cases.
```

Comment:
Agree that this is useful and not trying to block the implementation. However, I _do_ want to make sure the language in the proposal is accurate, and that it suggests the right path forward for people who come along later and think about webhook based validation.

It's also important to note that for most validation that people actually want to do this is going to be insufficient, since its going to be things like "field foo has to be a dns name" or "field bar is only present if field baz is not"

Again, perfectly happy to see the work here get done, just want to make sure we paint the right picture for what is to come in the future.